src/crimson/os/seastore/transaction.h

```diff
@@ -321,12 +321,14 @@ class Transaction {
       return false;
     }
     auto iter = retired_set.lower_bound(paddr);
-    auto &extent = iter->extent;
     if (iter == retired_set.end() ||
-	extent->get_paddr() > paddr) {
-      assert(iter != retired_set.begin());
+	iter->extent->get_paddr() > paddr) {
+      if (iter == retired_set.begin()) {
+	return false;
+      }
```

Comment:
> I'm not sure why is_retired(paddr, len) doesn't/cannot do an exact check, which can be simpler.

It can do the check, I just think searching through the retired_set for every check might be expensive in terms of cpu overhead.

Reply:
Not sure why doing an exact check is more expensive, it can still use lower_bound() but return false if paddr or len doesn't match.

This can also avoid the below issue/gap:
> If it is checking whether an arbitrary range is retired, it will be wrong if the range is retired but in two consecutive retired extents.

Reply:
> Not sure why doing an exact check is more expensive, it can still use lower_bound() but return false if paddr or len doesn't match.
> 
> This can also avoid the below issue/gap:
> 
> > If it is checking whether an arbitrary range is retired, it will be wrong if the range is retired but in two consecutive retired extents.

Sorry, I misunderstood your question, will change it to do exact checks.